### ISSUE TITLE
Resonance worker returns only duplicate memories instead of diverse anchors

### ISSUE THREAD

**Issue by leonvanbokhorst** [Research-Group-IxD/convai-narrative-memory-poc#18]
## Problem

The resonance worker is returning duplicate anchors instead of retrieving a diverse set of memories based on salience, time, and semantic relevance. This fundamentally breaks the intended behavior where multiple different memories should be woven together in retellings.

## Current Behavior

From `results/demo-three-retells-20251003T053153Z.json`:

**Seeded anchors:**
- 3 distinct memories (BB-8 calibration, Fontys demo, Rotterdam pilot)
- Different timestamps (12 hours, 14 days, 210 days ago)
- Varying salience (1.0, 0.95, 0.9)

**Retrieved beats:**
```json
"beats": [
  {
    "anchor_id": "e67e71ad-dba5-43ca-bd44-236eaa1423c9",
    "text": "While calibrating the narrative memory today, we watched BB-8 mimic the reteller's cadence.",
    "perceived_age": "yesterday",
    "activation": 0.62370795
  },
  {
    "anchor_id": "92def8f1-def1-4fef-8657-e73bb8990dde",
    "text": "While calibrating the narrative memory today, we watched BB-8 mimic the reteller's cadence.",
    "perceived_age": "yesterday",
    "activation": 0.62370795
  },
  {
    "anchor_id": "f2ab1425-2208-4e44-be6c-d2a3dd422c81",
    "text": "While calibrating the narrative memory today, we watched BB-8 mimic the reteller's cadence.",
    "perceived_age": "yesterday",
    "activation": 0.62370795
  }
]
```

**Issue:** All 3 beats are the SAME memory (identical text, same perceived age, same activation), but with different anchor IDs. The other two seeded memories (Fontys, Rotterdam) are completely absent.

## Expected Behavior

The resonance worker should:
1. Retrieve semantically diverse memories based on the query
2. Apply time decay correctly to differentiate recent vs. old memories
3. Weight by salience to prioritize important memories
4. Return a mix that reflects the actual memory landscape

**Expected output for the query "Retell the highlights from our narrative memory experiments":**
```json
"beats": [
  {
    "text": "While calibrating the narrative memory today, we watched BB-8 mimic...",
    "perceived_age": "yesterday",
    "activation": 0.998  // high: recent + high salience
  },
  {
    "text": "A couple of weeks back, we demoed the prototype at Fontys...",
    "perceived_age": "about 2 weeks ago",
    "activation": 0.924  // medium: some decay but high relevance
  },
  {
    "text": "Last autumn, in Rotterdam, the Reflective City pilot sparked...",
    "perceived_age": "about 7 months ago",
    "activation": 0.456  // lower: significant decay
  }
]
```

### Possible causes in `convai_narrative_memory_poc/workers/resonance/main.py`:

1. **Duplicate indexing:** The indexer may be storing the same memory multiple times (check `demo_three_retells.py` - are anchors being produced multiple times?)

2. **Search retrieval issue:** Qdrant search may not be retrieving diverse results:
   - Query embedding might be biased toward one anchor
   - `top_k=5` is requested but only 3 are returned, all duplicates
   - Collection may have duplicate entries from previous runs

3. **Activation calculation:** The activation formula may not be differentiating memories:
   ```python
   act = float(h.score) * decay * sal
   ```
   If all have identical scores and similar timestamps, they'd rank identically

4. **Sorting logic:** After calculating activation, only top 3 are selected:
   ```python
   scored.sort(key=lambda x: x[0], reverse=True)
   beats = [... for act, h in scored[:3] ...]
   ```
   If duplicates have identical activation, all get selected before diverse memories

## Investigation Steps

1. **Check Qdrant collection state:**
   ```bash
   docker compose run --rm tools python convai_narrative_memory_poc/tools/inspect_qdrant.py --limit 20
   ```
   - Are there duplicate anchor IDs?
   - Are all 3 distinct memories actually stored?

2. **Verify indexer behavior:**
   - Does `demo_three_retells.py` call `produce_anchor()` only once per memory?
   - Check indexer logs for duplicate processing

3. **Test query embedding:**
   - Is the query "Retell the highlights..." semantically matching all 3 memories equally?
   - Check Qdrant search scores before activation calculation

### Solution 1: Deduplicate at search level
```python
# In resonance/main.py, after search_anchors()
seen_texts = set()
unique_hits = []
for h in hits:
    text = h.payload['text']
    if text not in seen_texts:
        seen_texts.add(text)
        unique_hits.append(h)
hits = unique_hits
```

### Solution 2: Fix anchor production
If the issue is in `demo_three_retells.py` or seed scripts:
- Ensure each anchor is produced exactly once
- Add unique constraint or check before producing

### Solution 3: Improve diversity in ranking
After activation calculation, enforce diversity:
```python
# Instead of just sorting by activation
# Cluster by text similarity and pick top from each cluster
from sklearn.metrics.pairwise import cosine_similarity

def diversify_beats(scored_beats, diversity_threshold=0.9):
    """Select diverse beats, avoiding near-duplicates"""
    selected = []
    for act, h in sorted(scored_beats, key=lambda x: x[0], reverse=True):
        # Check if too similar to already selected
        is_diverse = all(
            text_similarity(h.payload['text'], s.payload['text']) < diversity_threshold
            for _, s in selected
        )
        if is_diverse:
            selected.append((act, h))
        if len(selected) >= 3:
            break
    return selected
```

### Solution 4: Clear collection between runs
For testing:
```python
# In demo_three_retells.py, at start
from qdrant_client import QdrantClient
client = QdrantClient(url=os.getenv("QDRANT_URL", "http://localhost:6333"))
try:
    client.delete_collection("anchors")
except:
    pass
time.sleep(2)  # Wait for indexer to recreate
```

## Files to Investigate/Modify

1. `convai_narrative_memory_poc/workers/resonance/main.py`
   - Add deduplication logic
   - Improve diversity in beat selection
   - Add logging for activation scores and text snippets

2. `convai_narrative_memory_poc/tools/demo_three_retells.py`
   - Verify anchor production is not duplicating
   - Add collection cleanup at start

3. `convai_narrative_memory_poc/workers/indexer/main.py`
   - Check if duplicate anchors are being processed
   - Ensure anchor_id is truly unique

## Success Criteria

- Running `demo_three_retells.py` returns 3 DISTINCT memories in beats
- Activation scores reflect proper time decay (recent > old)
- Text content varies across all returned beats
- Retelling combines multiple memories, not just one repeated
- System handles edge cases (duplicate submissions) gracefully
